File: src/main/java/dev/vrba/secrethitler/engine/Election.java
package dev.vrba.secrethitler.engine;

import org.jetbrains.annotations.NotNull;

import java.util.Collection;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

public class Election {

    private final Map<UUID, Vote> votes;

    public Election(@NotNull Collection<UUID> players) {
        this.votes = players.stream().collect(
                Collectors.toMap(
                        player -> player,
                        player -> Vote.NONE
                )
        );
    }

    public @NotNull Election vote(@NotNull UUID player, @NotNull Vote vote) {
        if (!this.votes.containsKey(player)) {
            throw new IllegalArgumentException("Player is not present within the current election.");
        }

        this.votes.put(player, vote);
        return this;
    }

    public @NotNull Map<UUID, Vote> maskVotesUnlessFinished() {
        // There is no need for masking the votes if election is already finished
        if (this.isFinished())  {
            return this.votes;
        }

        // Otherwise, player should only know, whether a player has voted or not
        return this.votes.entrySet()
                .stream()
                .map(entry -> Map.entry(entry.getKey(), entry.getValue().mask()))
                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
    }

    public boolean isFinished() {
        return this.votes.values()
                .stream()
                .noneMatch(vote -> vote == Vote.NONE);
    }

    public boolean isSuccessful() {
        return this.votes.values()
                .stream()
                .mapToInt(Vote::getValue)
                .sum() > 0;
    }
}

File: src/main/kotlin/dev/vrba/secrethitler/engine/election/Election.kt
package dev.vrba.secrethitler.engine.election

import java.util.*

fun createElection(players: Collection<UUID>): Election = Election(players.associateWith { Vote.None })

class Election(private val votes: Map<UUID, Vote>) {

    val finished: Boolean = votes.values.none { it == Vote.None }

    val successful: Boolean = votes.values.sumOf { it.value } > 0

    val masked: Map<UUID, Vote> =
        // Only mask election if it is not finished yet
        if (finished) votes
        // When masked, all players should only distinguish, whether a given player has voted or not
        else votes.map { it.key to if (it.value == Vote.None) Vote.None else Vote.Hidden }.toMap()
}